File: partiql-ast/src/main/java/org/partiql/ast/v1/expr/Expr.java
package org.partiql.ast.v1.expr;

import org.jetbrains.annotations.NotNull;
import org.partiql.ast.v1.AstNode;
import org.partiql.ast.v1.AstVisitor;

/**
 * TODO docs, equals, hashcode
 */
public abstract class Expr extends AstNode {
    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        if (this instanceof ExprLit) {
            return visitor.visitExprLit((ExprLit) this, ctx);
        } else if (this instanceof ExprIon) {
            return visitor.visitExprIon((ExprIon) this, ctx);
        } else if (this instanceof ExprVar) {
            return visitor.visitExprVar((ExprVar) this, ctx);
        } else if (this instanceof ExprSessionAttribute) {
            return visitor.visitExprSessionAttribute((ExprSessionAttribute) this, ctx);
        } else if (this instanceof ExprPath) {
            return visitor.visitExprPath((ExprPath) this, ctx);
        } else if (this instanceof ExprCall) {
            return visitor.visitExprCall((ExprCall) this, ctx);
        } else if (this instanceof ExprParameter) {
            return visitor.visitExprParameter((ExprParameter) this, ctx);
        } else if (this instanceof ExprOperator) {
            return visitor.visitExprOperator((ExprOperator) this, ctx);
        } else if (this instanceof ExprNot) {
            return visitor.visitExprNot((ExprNot) this, ctx);
        } else if (this instanceof ExprAnd) {
            return visitor.visitExprAnd((ExprAnd) this, ctx);
        } else if (this instanceof ExprOr) {
            return visitor.visitExprOr((ExprOr) this, ctx);
        } else if (this instanceof ExprValues) {
            return visitor.visitExprValues((ExprValues) this, ctx);
        } else if (this instanceof ExprCollection) {
            return visitor.visitExprCollection((ExprCollection) this, ctx);
        } else if (this instanceof ExprStruct) {
            return visitor.visitExprStruct((ExprStruct) this, ctx);
        } else if (this instanceof ExprLike) {
            return visitor.visitExprLike((ExprLike) this, ctx);
        } else if (this instanceof ExprBetween) {
            return visitor.visitExprBetween((ExprBetween) this, ctx);
        } else if (this instanceof ExprInCollection) {
            return visitor.visitExprInCollection((ExprInCollection) this, ctx);
        } else if (this instanceof ExprIsType) {
            return visitor.visitExprIsType((ExprIsType) this, ctx);
        } else if (this instanceof ExprCase) {
            return visitor.visitExprCase((ExprCase) this, ctx);
        } else if (this instanceof ExprCoalesce) {
            return visitor.visitExprCoalesce((ExprCoalesce) this, ctx);
        } else if (this instanceof ExprNullIf) {
            return visitor.visitExprNullIf((ExprNullIf) this, ctx);
        } else if (this instanceof ExprSubstring) {
            return visitor.visitExprSubstring((ExprSubstring) this, ctx);
        } else if (this instanceof ExprPosition) {
            return visitor.visitExprPosition((ExprPosition) this, ctx);
        } else if (this instanceof ExprTrim) {
            return visitor.visitExprTrim((ExprTrim) this, ctx);
        } else if (this instanceof ExprOverlay) {
            return visitor.visitExprOverlay((ExprOverlay) this, ctx);
        } else if (this instanceof ExprExtract) {
            return visitor.visitExprExtract((ExprExtract) this, ctx);
        } else if (this instanceof ExprCast) {
            return visitor.visitExprCast((ExprCast) this, ctx);
        } else if (this instanceof ExprDateAdd) {
            return visitor.visitExprDateAdd((ExprDateAdd) this, ctx);
        } else if (this instanceof ExprDateDiff) {
            return visitor.visitExprDateDiff((ExprDateDiff) this, ctx);
        } else if (this instanceof ExprQuerySet) {
            return visitor.visitExprQuerySet((ExprQuerySet) this, ctx);
        } else if (this instanceof ExprMatch) {
            return visitor.visitExprMatch((ExprMatch) this, ctx);
        } else if (this instanceof ExprWindow) {
            return visitor.visitExprWindow((ExprWindow) this, ctx);
        } else {
            throw new IllegalStateException("Unexpected value: " + this);
        }
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/expr/Expr.kt
package org.partiql.ast.v1.expr

import org.partiql.ast.v1.AstNode
import org.partiql.ast.v1.AstVisitor

/**
 * TODO docs, equals, hashcode
 */
public abstract class Expr : AstNode() {
    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R = when (this) {
        is ExprLit -> visitor.visitExprLit(this, ctx)
        is ExprIon -> visitor.visitExprIon(this, ctx)
        is ExprVar -> visitor.visitExprVar(this, ctx)
        is ExprSessionAttribute -> visitor.visitExprSessionAttribute(this, ctx)
        is ExprPath -> visitor.visitExprPath(this, ctx)
        is ExprCall -> visitor.visitExprCall(this, ctx)
        is ExprParameter -> visitor.visitExprParameter(this, ctx)
        is ExprOperator -> visitor.visitExprOperator(this, ctx)
        is ExprNot -> visitor.visitExprNot(this, ctx)
        is ExprAnd -> visitor.visitExprAnd(this, ctx)
        is ExprOr -> visitor.visitExprOr(this, ctx)
        is ExprValues -> visitor.visitExprValues(this, ctx)
        is ExprCollection -> visitor.visitExprCollection(this, ctx)
        is ExprStruct -> visitor.visitExprStruct(this, ctx)
        is ExprLike -> visitor.visitExprLike(this, ctx)
        is ExprBetween -> visitor.visitExprBetween(this, ctx)
        is ExprInCollection -> visitor.visitExprInCollection(this, ctx)
        is ExprIsType -> visitor.visitExprIsType(this, ctx)
        is ExprCase -> visitor.visitExprCase(this, ctx)
        is ExprCoalesce -> visitor.visitExprCoalesce(this, ctx)
        is ExprNullIf -> visitor.visitExprNullIf(this, ctx)
        is ExprSubstring -> visitor.visitExprSubstring(this, ctx)
        is ExprPosition -> visitor.visitExprPosition(this, ctx)
        is ExprTrim -> visitor.visitExprTrim(this, ctx)
        is ExprOverlay -> visitor.visitExprOverlay(this, ctx)
        is ExprExtract -> visitor.visitExprExtract(this, ctx)
        is ExprCast -> visitor.visitExprCast(this, ctx)
        is ExprDateAdd -> visitor.visitExprDateAdd(this, ctx)
        is ExprDateDiff -> visitor.visitExprDateDiff(this, ctx)
        is ExprQuerySet -> visitor.visitExprQuerySet(this, ctx)
        is ExprMatch -> visitor.visitExprMatch(this, ctx)
        is ExprWindow -> visitor.visitExprWindow(this, ctx)
        else -> throw NotImplementedError()
    }
}
